src/Tutor.Core/DomainModel/KnowledgeComponents/KnowledgeComponentMastery.cs

```diff
@@ -54,6 +63,54 @@ public Result<AssessmentEvent> SelectSuitableAssessmentEvent(IAssessmentEventSel
             return assessmentEventSelector.SelectSuitableAssessmentEvent(KnowledgeComponent.Id, LearnerId);
         }
 
+        private void TryPass()
+        {
+            if (IsPassed)
+                return; 
+
+            if (Mastery >= 0.95)
+            {
+                Causes(new KnowledgeComponentPassed()
+                {
+                    KnowledgeComponentId = KnowledgeComponent.Id,
+                    LearnerId = LearnerId
+                });
+                TrySatisfy();
+            }
+        }
+
+        private void TryComplete(int aeIdForCurrentSubmission)
+        {
+            if (IsCompleted)
+                return; 
+
+            foreach (AssessmentEvent assessmentEvent in KnowledgeComponent.AssessmentEvents)
+            {
+                if (assessmentEvent.Id != aeIdForCurrentSubmission && assessmentEvent.Submissions.Count == 0)
+                    return;
+            }
+
+            Causes(new KnowledgeComponentCompleted()
+            {
+                KnowledgeComponentId = KnowledgeComponent.Id,
+                LearnerId = LearnerId
+            });
+            TrySatisfy();
```

Comment:
Why does TryComplete call TrySatisfy given that TrySatisfy only cares for IsPassed?

Reply:
Mostly because the concept of whether or not an AU (or KC in our case) is satisfied from cmi5 can depend on both it being Passed and it being Completed. While we are currently using an approach that uses the same 'moveOn' criteria for all KCs, we have discussed changing which one we're actually using pretty soon. So the idea was mostly to make that easier.
Actually, I feel like it might even be worth it to implement the concept of 'moveOn' criteria now, and a proper TrySatisfy method that works based on 'moveOn'. Even if we don't plan to do anything like connect different 'moveOn' values to KCs anytime soon (or ever), doing something like that would allow us to maybe put which one we're currently using into configuration instead of changing code.

Reply:
I added a globally configurable 'moveOn' criteria in https://github.com/Clean-CaDET/tutor/pull/24/commits/30388fda5f58516240066dc65cc35977ae617d70.